url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/net/http/HttpServer3xx.java

```diff
@@ -0,0 +1,68 @@
+package net.hetimatan.net.http;
+
+import java.io.IOException;
+
+
+
+import net.hetimatan.io.file.KyoroFile;
+import net.hetimatan.io.filen.ByteKyoroFile;
+import net.hetimatan.util.event.net.io.KyoroSocket;
+import net.hetimatan.util.http.HttpRequestHeader;
+import net.hetimatan.util.http.HttpRequest;
+import net.hetimatan.util.http.HttpResponse;
+import net.hetimatan.util.io.ByteArrayBuilder;
+import net.hetimatan.util.log.Log;
+
+public class HttpServer3xx extends HttpServerTemplate {
+
+	private static HttpServer3xx server = null;
+	public static void main(String[] args) {
+		server = new HttpServer3xx();
+		server.setPort(18080);
+		server.startServer(null);
+	}
+
+	@Override
+	public KyoroFile createResponse(HttpServerFront front, KyoroSocket socket, HttpRequest uri) throws IOException {
+		HttpHistory.get().pushMessage(sId+"#createResponse:"+front.sId+"\n");
+		String path = uri.getLine().getRequestURI().getPath();
+		int index = 0;
+		if(path.startsWith("http://")) {
+			index = path.indexOf("/", "http://".length());
+		}
+		String address = uri.getValue("mv");
+		path = path.substring(index);
+		if(path.startsWith("/301")) {
+			return createHeader(HttpResponse.STATUS_CODE_301_MOVE_PERMANENTLY, address);
+		}
+		if(path.startsWith("/302")) {
+			return createHeader(HttpResponse.STATUS_CODE_302_Found, address);
+		}
+		if(path.startsWith("/303")) {
+			return createHeader(HttpResponse.STATUS_CODE_303_SEE_OTHER, address);
+		}
+		if(path.startsWith("/307")) {
+			return createHeader(HttpResponse.STATUS_CODE_307_TEMPORARY_REDIRECT, address);
+		}
+
+		return super.createResponse(front, socket, uri);
+	}
+
+	public KyoroFile createHeader(String responce, String location) throws IOException {
+		if(Log.ON){Log.v(TAG, "HttpServer#createHeader");}
+		KyoroFile builder = new ByteKyoroFile();
+		try {
+			builder.addChunk(("HTTP/1.1 "+responce+"\r\n").getBytes());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;responce&#39; (index: 0 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;location&#39;. The same parameter type is validated here for example: url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/tool/SimpleHttpServer.java:99. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).